File: practice-2/solana-task-2/src/bin/create-token-metadata.rs
use dotenv::dotenv;

use solana_client::nonblocking::rpc_client::RpcClient;

use solana_sdk::{
    pubkey::Pubkey,
    transaction::Transaction,
    signer::keypair::{Keypair, Signer},
    commitment_config::CommitmentConfig,
};
use mpl_token_metadata::{
    programs::MPL_TOKEN_METADATA_ID,
    instructions::{CreateMetadataAccountV3, CreateMetadataAccountV3InstructionArgs, CreateMetadataAccountV3Builder},
    types::{DataV2},
};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    dotenv().ok();

    let keypair_base58 =
        std::env::var("SECRET_KEY").expect("SECRET_KEY must be set.");
    let keypair_bytes = bs58::decode(keypair_base58).into_vec()?;
    let signer_keypair = Keypair::from_bytes(&keypair_bytes)?;
    let mint_pubkey = Pubkey::from_str_const("8UReKEpeSN329LyAue2RjXPpv11hyQp9gpkEemHCH69F");
    
    let client = RpcClient::new_with_commitment(
        String::from("https://api.devnet.solana.com"),
        CommitmentConfig::confirmed(),
    );

    let (metadata_PDA, metadata_bump_seed) = Pubkey::find_program_address(
        &[b"metadata", MPL_TOKEN_METADATA_ID.as_ref(), mint_pubkey.as_ref()],
        &MPL_TOKEN_METADATA_ID
    );
    
    println!("PDA: {:?}", metadata_PDA);
    
    let data = DataV2 {
        name: "Not Not coin".to_string(),
        symbol: "NNC".to_string(),
        uri: "https://arweave.net/tQ0E2wu869poiv01OQGaMKMs9fHsl8HCKvAJqvRfLmU".to_string(),
        seller_fee_basis_points: 0,
        creators: None,
        collection: None,
        uses: None,
    };
    
    println!("data : {:?}", data);
    
    let args = CreateMetadataAccountV3InstructionArgs{
        data: data.clone(),
        is_mutable: true,
        collection_details: None,
    };
    
    println!("args : {:?}", args);


    let tx = CreateMetadataAccountV3Builder::new()
        .metadata(metadata_PDA)
        .mint(mint_pubkey)
        .mint_authority(signer_keypair.pubkey())
        .payer(signer_keypair.pubkey())
        .update_authority(signer_keypair.pubkey(), true)
        .system_program(solana_sdk::system_program::id())
        .rent(None)
        .data(data)
        .is_mutable(true)
        .instruction();


    let mut transaction = Transaction::new_signed_with_payer(
        &[tx],
        Some(&signer_keypair.pubkey()),
        &[&signer_keypair],
        client.get_latest_blockhash().await?,
    );

    let signature = client.send_and_confirm_transaction(&transaction).await?;
    println!("Transaction Signature: {}", signature);


    Ok(())
}

File: practice-2/solana-task-2-ts/create-token-metadata.ts
import 'dotenv/config';

import {
	Keypair,
	PublicKey,
	Connection,
	clusterApiUrl,
	Transaction,
	sendAndConfirmTransaction,
} from '@solana/web3.js';
import { createCreateMetadataAccountV3Instruction } from '@metaplex-foundation/mpl-token-metadata';

const TOKEN_METADATA_PROGRAM_ID = new PublicKey(
	'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s',
);
const tokenMint = new PublicKey('8FnWaniphsmh4rVkxsgHXVX4djZQxoxbhowewoNFHheq');

const SECRET_KEY = 'SECRET_KEY';
const privateKey = process.env[SECRET_KEY];

if (privateKey === undefined) {
	console.log(`Add ${SECRET_KEY} to .env!`);
	process.exit(1);
}

const connection = new Connection(clusterApiUrl('devnet'));
const signer = Keypair.fromSecretKey(new Uint8Array(JSON.parse(privateKey)));

const metadata = {
	name: 'Kibish Token',
	symbol: 'KBSHT',
	uri: 'https://maroon-abundant-boar-405.mypinata.cloud/ipfs/bafybeiggphv5m3rggvcmbqhqvaebrxk6ln4ak5eklj7ex4rjt6v736w6ii',
	sellerFeeBasisPoints: 0,
	creators: null,
	collection: null,
	uses: null,
};

const [metadataPDA, _metadataBump] = PublicKey.findProgramAddressSync(
	[
		Buffer.from('metadata'),
		TOKEN_METADATA_PROGRAM_ID.toBuffer(),
		tokenMint.toBuffer(),
	],
	TOKEN_METADATA_PROGRAM_ID,
);

const tx = new Transaction().add(
	createCreateMetadataAccountV3Instruction(
		{
			metadata: metadataPDA,
			mint: tokenMint,
			mintAuthority: signer.publicKey,
			payer: signer.publicKey,
			updateAuthority: signer.publicKey,
		},
		{
			createMetadataAccountArgsV3: {
				collectionDetails: null,
				data: metadata,
				isMutable: true,
			},
		},
	),
);

tx.feePayer = signer.publicKey;
tx.recentBlockhash = (await connection.getLatestBlockhash()).blockhash;

const signature = await sendAndConfirmTransaction(connection, tx, [signer]);

console.log(`Transaction: ${signature}`);
